core/server/src/main/java/alluxio/worker/netty/DataServerWriteHandler.java

```diff
@@ -0,0 +1,337 @@
+  }
+
+  /**
+   * Validates the block write request.
+   *
+   * @param msg the block write request
+   * @return empty string if the request valid
+   */
+  private String validateRequest(RPCProtoMessage msg) {
+    Protocol.WriteRequest request = (Protocol.WriteRequest) msg.getMessage();
+    if (request.getId() != mRequest.mId) {
+      return "The Ids do not match.";
+    }
+    if (request.getOffset() != mPosToQueue) {
+      return String
+          .format("Offsets do not match [received: %d, expected: %d].", request.getOffset(),
+              mPosToQueue);
+    }
+    return "";
+  }
+
+  /**
+   * Writes an error block write response to the channel and closes the channel after that.
+   *
+   * @param channel the channel
+   */
+  private void replyError(Channel channel, Protocol.Status.Code code, String message, Throwable e) {
+    channel.writeAndFlush(RPCProtoMessage.createResponse(code, message, e, null))
+        .addListener(ChannelFutureListener.CLOSE);
+  }
+
+  /**
+   * Writes a response to signify the success of the block write. Also resets the channel.
+   *
+   * @param channel the channel
+   */
+  private void replySuccess(Channel channel) {
+    channel.writeAndFlush(RPCProtoMessage.createOkResponse(null))
+        .addListeners(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE, new ChannelFutureListener() {
+          @Override
+          public void operationComplete(ChannelFuture future) throws Exception {
+            reset();
+          }
+        });
+    channel.config().setAutoRead(true);
+    channel.read();
+  }
+
+  private void reset() throws IOException {
+    if (mRequest != null) {
+      mRequest.close();
+      mRequest = null;
+    }
+
+    mPosToQueue = 0;
+    mPosToWrite = 0;
+
+    try {
+      mLock.lock();
+      for (ByteBuf buf : mPackets) {
+        ReferenceCountUtil.release(buf);
+      }
+      mPacketWriterActive = false;
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  /**
+   * A runnable that polls from the packets queue and writes to the block worker.
+   */
+  private final class PacketWriter implements Runnable {
+    private ChannelHandlerContext mCtx;
+
+    /**
+     * Creates an instance of {@link PacketWriter}.
+     *
+     * @param ctx the netty channel handler context
+     */
+    PacketWriter(ChannelHandlerContext ctx) {
+      mCtx = ctx;
+    }
+
+    @Override
+    public void run() {
+      try {
+        runInternal();
+      } catch (Throwable e) {
+        LOG.error("Failed to run PacketWriter.", e);
+        throw e;
+      }
+    }
+
+    /**
+     * The actual implementation of the runnable.
+     */
+    private void runInternal() {
+      while (true) {
+        ByteBuf buf;
+        mLock.lock();
+        try {
+          buf = mPackets.poll();
+          if (buf == null) {
+            mPacketWriterActive = false;
+            break;
+          }
+          if (!tooManyPacketsInFlight()) {
+            mCtx.channel().config().setAutoRead(true);
```

Comment:
This writer thread is the only place that will re-enable the autoreading of the channel. However, there are situations where this thread can "die" and stop running, depending on the producer (netty thread) to start this thread again. However, the producer cannot start unless this writer thread enables the producer. Therefore, is there a situation where this thread does not have a chance to start the producer thread, and then dies, and never gets restarted, thus stalling the entire process?